File: domain/src/main/java/com/pardo/roomwithaword/dao/GenericDao.java
package com.pardo.roomwithaword.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Update;

import java.util.List;

@Dao
public interface GenericDao<T> {

    @Insert
    void insert(T data);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void persist(T data);

    @Delete
    void delete(T data);

    @Delete
    void delete(List<T> data);

    @Update
    void update(T data);

    @Update
    void update(List<T> data);

}

File: genericroom/src/main/java/com/github/fcopardo/room/base/GenericDao.kt
package com.github.fcopardo.room.base

import android.arch.persistence.room.Dao
import android.arch.persistence.room.Delete
import android.arch.persistence.room.Insert
import android.arch.persistence.room.OnConflictStrategy
import android.arch.persistence.room.Update

@Dao
interface GenericDao<T> {

    @Insert
    fun insert(data: T)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun persist(data: T)

    @Delete
    fun delete(data: T)

    @Delete
    fun delete(data: List<T>)

    @Update
    fun update(data: T)

    @Update
    fun update(data: List<T>)
}
